node/src/components/block_validator.rs

```diff
@@ -230,33 +230,36 @@ where
 
                 self.validation_states.retain(|key, state| {
                     if !state.missing_deploys.contains(&deploy_hash) {
-                        true
-                    } else {
-                        match state.source() {
-                            Some(peer) => {
-                                info!(%deploy_hash, ?peer, "trying the next peer");
-                                // There's still hope to download the deploy.
-                                let (chainspec, block_timestamp) = &state.context;
-                                effects.extend(
-                                    fetch_deploy(effect_builder,
-                                        Arc::clone(chainspec),
-                                        *block_timestamp,
-                                        deploy_hash,
-                                        peer,
-                                    ));
-                                retried = true;
-                                true
-                            },
-                            None => {
-                                // Notify everyone still waiting on it that all is lost.
-                                info!(block=?key, %deploy_hash, "could not validate the deploy. block is invalid");
-                                // This validation state contains a failed deploy hash, it can never
-                                // succeed.
-                                state.responders.drain(..).for_each(|responder| {
-                                    effects.extend(responder.respond((false, key.clone())).ignore());
-                                });
-                                false
-                            }
+                        return true
+                    }
+                    if retried {
+                        // We don't want to retry downloading the same element more than once.
```

Comment:
Not sure I'm following. So if we have three sources we'd only try two of them?

Reply:
If there are multiple blocks that depend on that particular deploy (multiple entries in `self.validation_states` where `state.missing_deploys.contains(&deploy_hash) = true`), we don't want to send a request to download that particular `deploy_hash` for each of them. It's enough to send one, and resolve all of the dependencies when it's downloaded.

Reply:
Ah, makes sense; I misunderstood, sorry! :+1: